File: Baekjoon/14888_연산자 끼워넣기/B_14888_2.js
const calc = function({idx, sums, p, s, m, d}) {
    if (idx === N - 1) {
        // -0은 0이 되도록 처리
        if (sums === -0) {
            sums = 0
        }

        if (maxNum < sums) {
            maxNum = sums
        }

        if (minNum > sums) {
            minNum = sums
        }
    } else {
        if (p > 0) {
            calc({
                idx: idx + 1,
                sums: sums + numbers[idx + 1],
                p: p - 1,
                s,
                m,
                d
            })
        }

        if (s > 0) {
            calc({
                idx: idx + 1,
                sums: sums - numbers[idx + 1],
                p,
                s: s - 1,
                m,
                d
            })
        }

        if (m > 0) {
            calc({
                idx: idx + 1,
                sums: sums * numbers[idx + 1],
                p,
                s,
                m: m - 1,
                d
            })
        }

        if (d > 0) {
            // 양수나 0이면 내림
            if (sums >= 0) {
                sums = Math.floor(sums / numbers[idx + 1])
            // 음수면 올림
            } else {
                sums = Math.ceil(sums / numbers[idx + 1])
            }
            calc({
                idx: idx + 1,
                sums,
                p,
                s,
                m,
                d: d - 1
            })
        }
    }
}

const fs = require('fs')
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n')

const N = Number(input[0])
const numbers = input[1].split(' ').map(v => Number(v))
const operators = input[2].split(' ').map(v => Number(v))

let maxNum = -Infinity
let minNum = Infinity

calc({
    idx: 0,
    sums: numbers[0],
    p: operators[0],
    s: operators[1],
    m: operators[2],
    d: operators[3]
})

console.log(maxNum)
console.log(minNum)

// 참고
// https://www.youtube.com/watch?v=Jz8Sx1XYb04
// 노마드코더의 clean code 관련 영상으로 인자가 많은 경우에는 object로 보내는 것을 추천한다는 내용에 영향을 받았다

File: baekjoon/14888_연산자 끼워넣기/B_14888_2.py
# B_14888 복사본 파일과의 차이를 잘 이해하기!
# idx 및 기타 요소들을 직접 수정해서 변수의 값 자체를 바꾸는 것과 인자의 값만 바꾸는 것의 차이
# 사용하고 해제를 해줘야 한다!

def back_track(idx, start_number, addition, subtraction, multiply, division):
    global max_num, min_num
    if idx == N - 1:
        max_num = max(start_number, max_num)
        min_num = min(start_number, min_num)
        return
    else:
        if addition:
            idx += 1
            start_number += numbers[idx]
            addition -= 1
            back_track(idx, start_number, addition, subtraction, multiply, division)
            start_number -= numbers[idx]
            idx -= 1
            addition += 1

        if subtraction:
            idx += 1
            start_number -= numbers[idx]
            subtraction -= 1
            back_track(idx, start_number, addition, subtraction, multiply, division)
            start_number += numbers[idx]
            idx -= 1
            subtraction += 1

        if multiply:
            idx += 1
            start_number *= numbers[idx]
            multiply -= 1
            back_track(idx, start_number, addition, subtraction, multiply, division)
            start_number //= numbers[idx]
            idx -= 1
            multiply += 1

        if division:
            if start_number >= 0:
                idx += 1
                start_number //= numbers[idx]
                division -= 1
                back_track(idx, start_number, addition, subtraction, multiply, division)
                start_number *= numbers[idx]
                idx -= 1
                division += 1
            else:
                idx += 1
                start_number *= -1
                start_number //= numbers[idx]
                start_number *= -1
                division -= 1
                back_track(idx, start_number, addition, subtraction, multiply, division)
                start_number *= numbers[idx]
                idx -= 1
                division += 1


N = int(input())
numbers = list(map(int, input().split()))
operators = list(map(int, input().split()))

max_num = -float('inf')
min_num = float('inf')

start = numbers[0]
back_track(0, start, operators[0], operators[1], operators[2], operators[3])
print(max_num)
print(min_num)
